src/lib/Type.hs

```diff
@@ -115,9 +115,31 @@ getClassTy (ClassDef _ _ bs _ _) = go bs
 
 -- === querying effects ===
 
-exprEffects :: (MonadFail1 m, EnvReader m) => Expr n -> m n (EffectRow n)
+computeAbsEffects :: (EnvExtender m, SubstE Name e)
+  => Abs (Nest Decl) e n -> m n (Abs (Nest Decl) (EffectRow `PairE` e) n)
+computeAbsEffects it = refreshAbs it \decls result -> do
+  effs <- declNestEffects decls
+  return $ Abs decls (effs `PairE` result)
+{-# INLINE computeAbsEffects #-}
+
+declNestEffects :: (EnvReader m) => Nest Decl n l -> m l (EffectRow l)
+declNestEffects decls = liftEnvReaderM $ declNestEffectsRec decls mempty
+{-# INLINE declNestEffects #-}
+
+declNestEffectsRec :: Nest Decl n l -> EffectRow l -> EnvReaderM l (EffectRow l)
+declNestEffectsRec Empty !acc = return acc
+declNestEffectsRec n@(Nest decl rest) !acc = withExtEvidence n $ do
```

Comment:
This is only a suggestion. This code is ok, but we have `BlockArguments` enabled, so `$ do ...` is equivalent to just `do ...`

Reply:
Done.